main.cpp

```diff
@@ -108,7 +108,410 @@ write 3 UDTs below that EACH have:
  */
 
 #include <iostream>
+#include <vector>
+#include <string>
+#include <unordered_set>
+
+struct Loan
+{
+    double loanAmount;
+    double interestRateInPercent;
+    unsigned long numPayments;
+    bool isSecured;
+    std::vector<double> payments;
+
+    Loan(double amount, double rate, unsigned long numPay, bool secured);
+    ~Loan();
+    void calculatePayments();
+    double calculateOverpayment();
+};
+//--
+Loan::Loan(double amount, double rate, unsigned long numPay, bool secured):
+    loanAmount(amount), interestRateInPercent(rate), numPayments(numPay), isSecured(secured)
+{
+    if(isSecured)
+    {
+        interestRateInPercent *= 0.9;
+    }
+    calculatePayments();
+}
+//--
+Loan::~Loan()
+{
+    std::cout << "Loan destructor called!" << std::endl;
+}
+//--
+void Loan::calculatePayments()
+{
+    payments.clear();
+    for(unsigned int i = 0; i < numPayments; i++)
+        {
+            double currentLoanAmount = loanAmount;
+            double payment = currentLoanAmount * (interestRateInPercent / 12 / 100.0) + loanAmount / static_cast<double>(numPayments);
+            currentLoanAmount -= loanAmount / static_cast<double>(numPayments);
+            payments.push_back(payment);
+        }
+}
+//--
+double Loan::calculateOverpayment()
+{
+    double amountToPay = 0.0;
+    for(unsigned int i = 0; i < numPayments; i++)
+        {
+            amountToPay += payments[i];
+        }
+    return amountToPay - loanAmount;
+}
+//--
+struct Gym 
+{
+    struct Member 
+    {
+        std::string name;
+        static int memberId;
+        int age;
+        bool isActive = true;
+        std::vector<std::string> activityLog;
+
+        Member(std::string n, int a);
+        ~Member();
+        void logActivity(const std::string& activity);
+        void printDetails();
+    };
+
+    std::vector<Member> members;
+    std::string gymName;
+    int totalMembers = 0;
+    double membershipPrice = 30.0;
+    double monthlyRevenue = 0.0;
+    bool isOpen;
+
+    Gym(std::string name, bool open);
+    ~Gym();
+    void registerMember(const Member& member);
+    void toggleGymStatus();
+    void calculateRevenue();
+};
+//--
+int Gym::Member::memberId = 0;
+//--
+Gym::Member::Member(std::string n, int a):
+    name(n), age(a)
+{
+    memberId++;
+}
```

Comment:
Everywhere that you have `i++`, `i--`, `i -= 1;` or `i += 1;`, you should use `pre-increment/decrement`. 
You are currently doing `post-increment/decrement`.  
you can think of post- and pre- increment like this:
```
int preIncrement(int& i)
{
    i += 1;
    return i; //both the returned value AND i have changed
}

int postIncrement(int& i)
{
    int oldVal = i;
    i += 1;
    return oldVal; //i has changed but the value being returned does not reflect this change.
}
```
more information can be found here explaining why `pre-increment/decrement` is preferred over `post-increment/decrement`: https://en.cppreference.com/w/cpp/language/operator_incdec
- [ ] revise

- [ ] revise